usecases/ingested_data_reader_usecase.go

```diff
@@ -2,34 +2,294 @@ package usecases
+		uniqueFieldName string,
+	) ([]models.DataModelObject, error)
+}
+
+type ingestedDataReaderRepository interface {
+	GetDataModel(
+		ctx context.Context,
+		exec repositories.Executor,
+		organizationID string,
+		fetchEnumValues bool,
+	) (models.DataModel, error)
+	ListPivots(
+		ctx context.Context,
+		exec repositories.Executor,
+		organization_id string,
+		tableId *string,
+	) ([]models.PivotMetadata, error)
+}
+
+type ingestedDataReaderIndexReader interface {
+	ListAllUniqueIndexes(ctx context.Context, organizationId string) ([]models.UnicityIndex, error)
+}
+
 type IngestedDataReaderUsecase struct {
-	ingestedDataReadRepository repositories.IngestedDataReadRepository
-	dataModelRepository        repositories.DataModelRepository
-	executorFactory            executor_factory.ExecutorFactory
+	clientDbRepository            ingestedDataReaderClientDbRepository
+	repository                    ingestedDataReaderRepository
+	executorFactory               executor_factory.ExecutorFactory
+	ingestedDataReaderIndexReader ingestedDataReaderIndexReader
+}
+
+func NewIngestedDataReaderUsecase(
+	clientDbRepository ingestedDataReaderClientDbRepository,
+	repository ingestedDataReaderRepository,
+	executorFactory executor_factory.ExecutorFactory,
+	ingestedDataReaderIndexReader ingestedDataReaderIndexReader,
+) IngestedDataReaderUsecase {
+	return IngestedDataReaderUsecase{
+		clientDbRepository:            clientDbRepository,
+		repository:                    repository,
+		executorFactory:               executorFactory,
+		ingestedDataReaderIndexReader: ingestedDataReaderIndexReader,
+	}
 }
 
-func (usecase *IngestedDataReaderUsecase) GetIngestedObject(ctx context.Context,
-	organizationId string, objectType string, objectId string,
-) ([]models.DataModelObject, error) {
+func (usecase IngestedDataReaderUsecase) GetIngestedObject(
+	ctx context.Context,
+	organizationId string,
+	dataModel *models.DataModel,
+	objectType string,
+	uniqueFieldValue string,
+	uniqueFieldName string,
+) ([]models.ClientObjectDetail, error) {
+	if dataModel == nil {
+		d, err := usecase.repository.GetDataModel(ctx,
+			usecase.executorFactory.NewExecutor(), organizationId, false)
+		if err != nil {
+			return nil, err
+		}
+		dataModel = &d
+	}
+
+	table := dataModel.Tables[objectType]
+
+	db, err := usecase.executorFactory.NewClientDbExecutor(ctx, organizationId)
+	if err != nil {
+		return nil, err
+	}
+
+	objects, err := usecase.clientDbRepository.QueryIngestedObjectByUniqueField(ctx, db, table, uniqueFieldValue, uniqueFieldName)
+	if err != nil {
+		return nil, err
+	}
+
+	clientObjects := make([]models.ClientObjectDetail, 0, len(objects))
+	for _, object := range objects {
+		validFrom, _ := object.Metadata["valid_from"].(time.Time)
+		clientObject := models.ClientObjectDetail{
+			Data:     object.Data,
+			Metadata: models.ClientObjectMetadata{ValidFrom: validFrom, ObjectType: objectType},
+		}
+		clientObjects = append(clientObjects, clientObject)
+	}
```

Comment:
Same remark as above, if you know beforehand how many items will be in the array (and it does not vary depending on a condition in the loop body), use indexing instead of append.

_(I know, micro-optimizations 🫣)_